src/pip/_internal/req/req_install.py

```diff
@@ -150,6 +150,8 @@ def __init__(
         self.global_options = global_options if global_options else []
         self.hash_options = hash_options if hash_options else {}
         self.config_settings = config_settings
+        if isinstance(comes_from, InstallRequirement) and comes_from.config_settings:
+            self.config_settings = comes_from.config_settings
```

Comment:
Does this handle the case if a package’s dependency specifies a config setting? Say

```
flask --config-settings=foo:1

# A dependency of flask
jinja2 --config-settings=foo:2
```

What config settings does Jinja2 receive? Does ordering matter?

Reply:
it always returns foo=2 in ArgRecordingSdist.args()

Reply:
maybe because of https://github.com/pypa/setuptools/issues/1928

Reply:
but in your example comes_from is str, not InstallRequirement

Reply:
So this means the `isinstance(comes_from, InstallRequirement)` case is only possible if `self` is _not_ directly requested by the user, but a transitive dependency of another package. So

1. If a user-requested package has config settings, those are always used.
2. If a user-requested package does not have user-specified config settings, but is another package’s transitive dependency, it would inherit the dependnt’s config settings (if any)
3. A transitive cannot have user-specified config settings.
4. If a transitive dependency’s dependant has config settings, the config settings are inherited.

Am I understanding this correctly? It’d probably be a good idea to add some comments in the source so it’s easier to analyse all cases, and spot potential bugs if some of the assumptions change in the future (i.e. if a transitive dependency’s `comes_from` somehow is an `InstallRequirement`).

Reply:
Yes, you are correct